sdk/core/Azure.Core/src/Internal/ArrayBackedPropertyBag.cs

```diff
@@ -260,35 +258,51 @@ private int GetIndex(TKey key)
             if (_count > 1 && _second.Key.Equals(key))
                 return 1;
 
+            if (_count <= 2)
+                return -1;
+
+            Kvp[] rest = GetRest();
             int max = _count - 2;
             for (var i = 0; i < max; i++)
             {
-                if (_rest[i].Key.Equals(key))
+                if (rest[i].Key.Equals(key))
                     return i + 2;
             }
             return -1;
         }
 
         internal void Dispose()
         {
-            CheckDisposed();
 #if DEBUG
+            if (_disposed)
```

Comment:
do we expect this to be thread safe? If yes, I think we need to add some memory barrier otherwise we can still dispose multiple times.

Reply:
No, it is not thread-safe, and neither `PipelineRequest` or `HttpMessage`.